src/checkpoints/checkpoints.cpp

```diff
@@ -143,97 +145,35 @@ namespace cryptonote
-                         false,
-                         "Vote is indexing out of bounds");
-
-    const auto signature_it = std::find_if(checkpoint.signatures.begin(), checkpoint.signatures.end(), [&vote](service_nodes::voter_to_signature const &check) {
-      return vote.voters_quorum_index == check.quorum_index;
-    });
-
-    if (signature_it == checkpoint.signatures.end())
-    {
-      service_nodes::voter_to_signature new_voter_to_signature = {};
-      new_voter_to_signature.quorum_index                      = vote.voters_quorum_index;
-      new_voter_to_signature.signature                         = vote.signature;
-      checkpoint.signatures.push_back(new_voter_to_signature);
-      return true;
-    }
-
-    return false;
-  }
-  //---------------------------------------------------------------------------
-  bool checkpoints::add_checkpoint_vote(service_nodes::checkpoint_vote const &vote)
-  {
-   // TODO(doyle): Always accept votes. Later on, we should only accept votes up
-   // to a certain point, so that eventually once a certain number of blocks
-   // have transpired, we can go through all our "seen" votes and deregister
-   // anyone who has not participated in voting by that point.
-
-#if 0
-    uint64_t newest_checkpoint_height = get_max_height();
-    if (vote.block_height < newest_checkpoint_height)
-      return true;
-#endif
-
-    // TODO(doyle): Double work. Factor into a generic vote checker as we already have one in service node deregister
-    std::array<int, service_nodes::QUORUM_SIZE> unique_vote_set = {};
-    std::vector<checkpoint_t> &candidate_checkpoints            = m_staging_points[vote.block_height];
-    std::vector<checkpoint_t>::iterator curr_checkpoint         = candidate_checkpoints.end();
-    for (auto it = candidate_checkpoints.begin(); it != candidate_checkpoints.end(); it++)
+    // TODO(doyle): Verify signatures and hash check out
+    std::array<size_t, service_nodes::CHECKPOINT_QUORUM_SIZE> unique_vote_set = {};
+    if (checkpoint.type == checkpoint_type::service_node)
     {
-      checkpoint_t const &checkpoint = *it;
-      if (checkpoint.block_hash == vote.block_hash)
-        curr_checkpoint = it;
-
+      CHECK_AND_ASSERT_MES(checkpoint.signatures.size() >= service_nodes::CHECKPOINT_MIN_VOTES, false, "Checkpoint has insufficient signatures to be considered");
       for (service_nodes::voter_to_signature const &vote_to_sig : checkpoint.signatures)
       {
-        if (vote_to_sig.quorum_index > unique_vote_set.size())
-          return false;
-
-        if (++unique_vote_set[vote_to_sig.quorum_index] > 1)
-        {
-          // NOTE: Voter is trying to vote twice
-          return false;
-        }
+        CHECK_AND_ASSERT_MES(vote_to_sig.voter_index < service_nodes::CHECKPOINT_QUORUM_SIZE, false, "Vote is indexing out of bounds");
+        CHECK_AND_ASSERT_MES(++unique_vote_set[vote_to_sig.voter_index] >= 1, false, "Voter is trying to vote twice");
```

Comment:
It is assert, so should be == 1, or <= 1?

Reply:
The `++` side effect in a macro argument is a bit terrifying (and something similar to this--a side effect in an `assert()`--was the cause of a lokinet bug for a while that broke non-debug builds).

Reply:
Yes it should be ==1 or <=1, have updated and moved the side effect out of the macro.